### PR TITLE
Simplex noise for foggies

### PR DESCRIPTION
**Pull request by Paril**
Simplex noise. Maps will need slight adjustments to look nice since this is an entirely different algorithm.
Tokays might need a higher frequency, whereas Edge may need amplitude, octave or noise scale adjustments.

### CODE REVIEW

File: src/client/renderer/r_bsp_draw.c [hunk 1]
```diff
@@ -170,17 +170,21 @@ void R_DrawBspLightgrid(r_view_t *view) {
...
+						const byte b = Mini(fog[2], 255);
+						const float af = a / 255.f;
+
+						R_AddSprite(view, &(r_sprite_t) {
+							.origin = origin,
+							.size = 8.f,
+							.color = Color32(r * af, g * af, b * af, 0),
+							.media = (r_media_t *) particle
+						});
+					}
```
Comment (jdolan) [inline src/client/renderer/r_bsp_draw.c:187]: Smart.

File: src/tools/quemap/fog.c [hunk 2]
```diff
@@ -66,7 +66,12 @@ static void FogForEntity(const cm_entity_t *entity) {
 		if (Cm_EntityValue(entity, "fog_absorption")->parsed ||
 			Cm_EntityValue(entity, "fog_color")->parsed ||
 			Cm_EntityValue(entity, "fog_density")->parsed ||
-			Cm_EntityValue(entity, "fog_noise")->parsed) {
+			Cm_EntityValue(entity, "fog_noise")->parsed ||
+			Cm_EntityValue(entity, "fog_frequency")->parsed ||
+			Cm_EntityValue(entity, "fog_amplitude")->parsed ||
+			Cm_EntityValue(entity, "fog_lacunarity")->parsed ||
```
Comment (jdolan) [inline src/tools/quemap/fog.c:72]: I've never heard this word before.

Side note, you like the entity parsing stuffs, doncha? :)

File: src/tools/quemap/lightgrid.c [hunk 3]
```diff
@@ -433,6 +433,176 @@ void IndirectLightgrid(int32_t luxel_num) {
...
+        output += (amplitude * SimplexNoise(x * frequency, y * frequency, z * frequency));
+        denom += amplitude;
+
+        frequency *= lacunarity;
+        amplitude *= persistence;
+    }
+
+    return (output / denom);
+}
+
```
Comment (jdolan) [inline src/tools/quemap/lightgrid.c:605]: Could we move all of this to its own file, like what we did for the ambient occlusion shapes? Just to keep the lightgrid code a bit more focused / readable.

**Review by jdolan — APPROVED** [review]
Code looks awesome. Thank you for adhering to all of my documentation and organization peculiarities. I'm excited to play with this! We'll also want to update `target/default/scripts/entities.def` in quetoo-data when / as we merge this in.

File: src/tools/quemap/lightgrid.c (hunk 3 above)
Comment (Paril) [reply src/tools/quemap/lightgrid.c]: Done. Simplex.h/simplex.c in Quemap. No other lib will use it I don't think so I just isolated it to there.

File: src/tools/quemap/lightgrid.c [hunk 4]
```diff
@@ -453,7 +454,18 @@ static void FogLuxel(GArray *fogs, luxel_t *l, float scale) {
 				break;
 		}
 
-		intensity *= fog->density + RandomRangef(-fog->noise, fog->noise);
+		// TODO: Should the divisors below be lg->w instead of MAX_BSP_LIGHTGRID_WIDTH?
```
Comment (jdolan) [inline src/tools/quemap/lightgrid.c:457]: @Paril This is my one remaining question on this one before we merge in. This looks really nice!

**Review by jdolan — APPROVED** [review]

**Pull request merged**